### PR TITLE
Fix issues with json-c v0.13 (Fixes #1355)

### LINKED ISSUE #1355: Problems with git master versions of json-c
Short version:  It won't compile out of the box.  If do you change the code in order to compile it, it will crash whenever you use any key binding.  I eventually decided to see if there were earlier tags of json-c that wouldn't have this problem and it seems that even with the latest one, version 0.12.1, sway compiles and runs just fine.  Maybe the maximum version of json-c with which sway will work should be indicated in README.

Long version:
When compiling, you get type mismatch errors in five files, all involving json_object_array_length in for loops:
`
sway/ipc-server.c: In function ‘ipc_client_handle_command’:
sway/ipc-server.c:449:21: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
   for (int i = 0; i < json_object_array_length(request); i++) {
`

The five files are:
* sway/ipc-server.c
* swaybar/status_line.c
* swaybar/tray/tray.c
* swaygrab/json.c
* swaylock/main.c

The solution for that is very simple.  Merely change the type of the counter variables from int to size_t.

On pressing keys, I've traced the problem to the line in ipc-server.c:
```c
json_object_object_add(obj, "binding", sb_obj);
```

My guess is that the call to `json_object_put(obj);` later in the function frees the objects recursively and then sb_obj is used (or more likely freed, as the stdout error I get from glibc is `*** Error in 'sway': corrupted double-linked list: `) later, causing an Earth-shattering kaboom.

Here's the log file.  I used Lynx because I couldn't figure out how to upload a file with my usual browser (though, I know, Lynx should be be usual browser; with Lynx, all I had to do was ^Xe:e ~/.sway.log&lt;Return&gt;VGy:e \#1&lt;Return&gt;p:wq) and I seem not to have been logged into to github, so it's anonymous gist.  Hmm.
 https://gist.github.com/anonymous/fe18bb44c7d40d5d63fbf15427e70fdc

### PR DESCRIPTION
**Pull request by besser82**
This superseeds #1439 and fixes #1355.

### CODE REVIEW

**Comment by besser82** [comment]
This superseeds https://github.com/swaywm/sway/pull/1439.

File: sway/ipc-server.c [hunk 1]
```diff
@@ -1126,7 +1126,12 @@ static void ipc_event_binding(json_object *sb_obj) {
 	sway_log(L_DEBUG, "Sending binding::run event");
 	json_object *obj = json_object_new_object();
 	json_object_object_add(obj, "change", json_object_new_string("run"));
-	json_object_object_add(obj, "binding", sb_obj);
+	// When adding a referenced json_object with an unknown lifetime
+	// to another json_object, it must be done with a wrapped call to
+	// json_object_get().  Otherwise the object's refcount does not
+	// get properly incremented and it will be deleted together with
+	// the object it got added to.
```
Comment (SirCmpwn) [inline sway/ipc-server.c]: This comment belongs in the commit message

**Comment by SirCmpwn** [comment]
I hate that we have to do this terrible workaround, but I don't see a better way. This sucks.

**Comment by besser82** [comment]
@SirCmpwn This is actually *NOT* a workaround, but well documented behaviour of json_object_object_add() in the json-c API…

See:  https://github.com/json-c/json-c/blob/master/json_object.h#L379

```
Remember, when using
json_object_object_add or json_object_array_put_idx, ownership will
transfer to the object/array.  Call json_object_get if you want to maintain
shared ownership or also add this object as a child of multiple objects or
arrays.  Any ownerships you acquired but did not transfer must be released
through json_object_put.
```

and:  https://github.com/json-c/json-c/blob/master/json_object.h#L408

```
The reference count will *not* be incremented. This is to make adding
fields to objects in code more compact. If you want to retain a reference
to an added object, independent of the lifetime of obj, you must wrap the
passed object with json_object_get.

Upon calling this, the ownership of val transfers to obj.  Thus you must
make sure that you do in fact have ownership over this object.
```

**Comment by SirCmpwn** [comment]
I'm referring to the typedef and the version check header, not the refcounting.

**Comment by besser82** [comment]
Ahh, I see…  Well, the typedef is not needed, if we remove `-Werror`…  Converting from size_t to int is safe in all cases, since there are no json_array_objects bigger than 4 GiB used somewhere…

**Comment by SirCmpwn** [comment]
Nah, I prefer this to dropping Werror, since Werror is useful beyond this one json-c problem.

**Comment by SirCmpwn** [comment]
This has been merged into the 0.15 branch and will make the next bugfix release. Thanks!

**Comment by besser82** [comment]
@SirCmpwn What about the master-branch?

**Comment by SirCmpwn** [comment]
Don't worry about it.

**Pull request closed without merging**